lib/hexpm/accounts/web_auth.ex

```diff
@@ -0,0 +1,125 @@
+defmodule Hexpm.Accounts.WebAuth do
+  use Hexpm.Context
+
+  @moduledoc false
+
+  alias Hexpm.Accounts.WebAuthRequest
+
+  # `device_code` refers to the code assigned to a client to identify it
+  # `user_code` refers to the code the user enters to authorize a client
+  # `name` refers to the name given to a key
+  # `scopes` refers to the list of scopes that are allowed in a web auth request
+  # `key_permission` is the default permissions given to generate a hex api key
+  # `key_params` is the default params to generate a hex api key
+
+  @key_permission %{domain: "api", resource: nil}
+  @key_params %{name: nil, permissions: [@key_permission]}
+
+  @doc """
+  Adds a web auth request to the pool and returns the response.
+
+  ## Params
+
+    - `name` - Name of the keys to be generated.
+  """
+
+  def get_code(key_name) do
+    device_code =
+      32
+      |> :crypto.strong_rand_bytes()
+      |> Base.url_encode64(padding: false)
+
+    user_code =
+      5
+      |> :crypto.strong_rand_bytes()
+      |> Base.hex_encode32(padding: false)
+      |> String.split_at(4)
+      |> then(fn {x, y} -> x <> "-" <> y end)
+
+    params = %{
+      device_code: device_code,
+      user_code: user_code,
+      key_name: key_name
+    }
+
+    changeset = WebAuthRequest.create(%WebAuthRequest{}, params)
+
+    with {:ok, _} <- Repo.insert(changeset) do
+      {:ok, %{device_code: device_code, user_code: user_code}}
+    end
+  end
+
+  @doc """
+  Submits a verification request to the pool and returns the response.
+
+  ## Params
+
+    - `user` - The user for whom the key should be generated.
+    - `user_code` - The user code entered by that user.
+  """
+  def submit(user, user_code) do
+    request =
+      WebAuthRequest
+      |> Repo.get_by(user_code: user_code)
```

Comment:
Old requests should ideally be deleted on a daily basis, but I am still looking into how schedule like that. Any suggestions?

Reply:
You can add the clean up as a release task here: https://github.com/hexpm/hexpm/tree/main/lib/hexpm/release_tasks and we will add a nightly job for it.

But even with the release task I would like a time filter on the query as an extra assurance in case there are any issues with the release task.